File: Salary_Data_Naive_Bayes.py
##############################################################################
########################## Naive-Bayes #######################################
##############################################################################


#importing packages and loading the data

import pandas as pd
import numpy as np
salary_train = pd.read_csv("C:\\Users\\home\\Desktop\\Data Science Assignments\\Python_codes\\Naive Bayes\\SalaryData_Train.csv")
salary_test = pd.read_csv("C:\\Users\\home\\Desktop\\Data Science Assignments\\Python_codes\\Naive Bayes\\SalaryData_Test.csv")

#our targated variable is salary

salary_train.info()
salary_train.dtypes

#creating a list of columns which are objects excluding the "Salary" column
string_columns = ["workclass","education","maritalstatus","occupation","relationship","race","sex","native"]

##Preprocessing the data. As, there are categorical variables
from sklearn.preprocessing import LabelEncoder
#labelling the categorical columns
number = LabelEncoder()
for i in string_columns:
    salary_train[i]= number.fit_transform(salary_train[i])
    salary_test[i]=number.fit_transform(salary_test[i])
    


##Capturing the column names which can help in futher process
colnames = salary_train.columns
colnames
len(colnames)#14






######Splitting the train dta into feature set and tagreted set
x_train = salary_train[colnames[0:13]]
y_train = salary_train[colnames[13]]

##### Splitting the test dat into feature set and targated set
x_test = salary_test[colnames[0:13]]
y_test = salary_test[colnames[13]]





#############################################################
############## MODEL BUILDING ###############################
#############################################################

##Building Multinomial Naive Bayes model
from sklearn.naive_bayes import MultinomialNB as MB
from sklearn.naive_bayes import GaussianNB as GB

########  Building the Multinomial naive bayes model #########

classifier_mb = MB()
classifier_mb.fit(x_train,y_train)
pred_mb = classifier_mb.predict(x_train)
accuracy_mb_train = np.mean(pred_mb == y_train)#0.7729186698053778
#77.2%
pd.crosstab(pred_mb, y_train)
#Salary   <=50K   >50K
#row_0                
# <=50K   21717   5913
# >50K      936   1595


##predicting on test data
pred_mb_test = classifier_mb.predict(x_test)
accuracy_mb_test = np.mean(pred_mb_test == y_test)#0.7749667994687915
#77.4%
pd.crosstab(pred_mb_test, y_test)
#Salary   <=50K   >50K
#row_0                
# <=50K   10891   2920
# >50K      469    780



############## Building Gaussian model #####################

classifier_gb = GB()
classifier_gb.fit(x_train, y_train)
pred_gb = classifier_gb.predict(x_train)
accuracy_gb_train = np.mean(pred_gb == y_train)# 0.7953317197705646
#79.5%
pd.crosstab(pred_gb,y_train)
#Salary   <=50K   >50K
#row_0                
# <=50K   21505   5025
# >50K     1148   2483

##for test data
pred_gb_test = classifier_gb.predict(x_test)
accuracy_gb_test = np.mean(pred_gb_test == y_test)#0.7946879150066402
#79.4%
pd.crosstab(pred_gb_test,y_test)
#Salary   <=50K   >50K
#row_0                
# <=50K   10759   2491
# >50K      601   1209

File: Salary_Data_Naive_Bayes.R
############### Packages Required ##############

install.packages("naivebayes")
library(naivebayes)

library(ggplot2)

library(caret)

library(psych)

library(e1071)



########## Reading and understanding the data #########


####reading the train dataset #####
train_sal <- read.csv(file.choose())
View(train_sal)
nrow(train_sal)#[1] 30161
ncol(train_sal)#[1] 14
colnames(train_sal)
str(train_sal)

train_sal$educationno <- as.factor(train_sal$educationno)

class(train_sal)


####reading the test dataset ###########

test_sal <- read.csv(file.choose())

str(test_sal)

View(test_sal)
nrow(test_sal)#[1] 15060
ncol(test_sal)#[1] 14

test_sal$educationno <- as.factor(test_sal$educationno)

class(test_sal)

###################### Visualization ##############

##### Plot and ggplot #####

ggplot(data=train_sal,aes(x=train_sal$Salary, y = train_sal$age, fill = train_sal$Salary)) +
  geom_boxplot() +
  ggtitle("Box Plot")


plot(train_sal$workclass,train_sal$Salary)


plot(train_sal$education,train_sal$Salary)


plot(train_sal$educationno,train_sal$Salary)


plot(train_sal$maritalstatus,train_sal$Salary)


plot(train_sal$occupation,train_sal$Salary)


plot(train_sal$relationship,train_sal$Salary)


plot(train_sal$race,train_sal$Salary)


plot(train_sal$sex,train_sal$Salary)


ggplot(data=train_sal,aes(x=train_sal$Salary, y = train_sal$capitalgain, fill = train_sal$Salary)) +
  geom_boxplot() +
  ggtitle("Box Plot")


ggplot(data=train_sal,aes(x=train_sal$Salary, y = train_sal$capitalloss, fill = train_sal$Salary)) +
  geom_boxplot() +
  ggtitle("Box Plot")


ggplot(data=train_sal,aes(x=train_sal$Salary, y = train_sal$hoursperweek, fill = train_sal$Salary)) +
  geom_boxplot() +
  ggtitle("Box Plot")


plot(train_sal$native,train_sal$Salary)


########Density Plot ###############

ggplot(data=train_sal,aes(x = train_sal$age, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Age - Density Plot")

ggplot(data=train_sal,aes(x = train_sal$workclass, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Workclass Density Plot")

ggplot(data=train_sal,aes(x = train_sal$education, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')

ggtitle("education Density Plot")

ggplot(data=train_sal,aes(x = train_sal$educationno, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("educationno Density Plot")

ggplot(data=train_sal,aes(x = train_sal$maritalstatus, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("maritalstatus Density Plot")

ggplot(data=train_sal,aes(x = train_sal$occupation, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')



ggtitle("occupation Density Plot")

ggplot(data=train_sal,aes(x = train_sal$sex, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("sex Density Plot")

ggplot(data=train_sal,aes(x = train_sal$relationship, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Relationship Density Plot")

ggplot(data=train_sal,aes(x = train_sal$race, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Race Density Plot")

ggplot(data=train_sal,aes(x = train_sal$capitalgain, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Capitalgain Density Plot")

ggplot(data=train_sal,aes(x = train_sal$capitalloss, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Capitalloss Density Plot")

ggplot(data=train_sal,aes(x = train_sal$hoursperweek, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')


ggtitle("Hoursperweek Density Plot")

ggplot(data=train_sal,aes(x = train_sal$native, fill = train_sal$Salary)) +
  geom_density(alpha = 0.9, color = 'Violet')

ggtitle("native Density Plot")



####################### Naive Bayes Model ################################

Model <- naiveBayes(train_sal$Salary ~ ., data = train_sal)
Model

#model prediction
Model_pred <- predict(Model,test_sal)
mean(Model_pred==test_sal$Salary)#[1] 0.8187251


## confusion matrix

confusionMatrix(Model_pred,test_sal$Salary)
####Accuracy : 0.8187
# accuracy is 81.87%